plugins/base/src/test/kotlin/model/FunctionsTest.kt

```diff
@@ -0,0 +1,180 @@
+                |package function
+                | /**
+                | * generic function
+                | */
+                |public fun <T : R, R> generic() {}
+                |class Test<T>() {}
+        """
+        ) {
+            with((this / "function" / "generic").assertIsInstance<Function>("generic")) {
+                name.assertEqual("generic", "Function name: ")
+                visibility.values.first().assertEqual(org.jetbrains.kotlin.descriptors.Visibilities.PUBLIC)
+                parameters.assertCount(0)
+                returnType.assertNotNull("returnType").constructorFqName.assertEqual("kotlin.Unit", "Return type: ")
+
+                typeParameters.assertCount(2, "Type parameters: ")
+                val T = typeParameters.find { it.name == "T" }.assertIsInstance<GenericType>("T")
+                val R = typeParameters.find { it.name == "R" }.assertIsInstance<GenericType>("R")
+
+                with(T) {
+                    name.assertEqual("T", "First type parameter name: ")
+                    bounds.assertCount(1, "First type parameter bounds: ")
+                }
+                with(R) {
+                    name.assertEqual("R", "Second type parameter name: ")
+                    bounds.assertCount(1, "Second type parameter bounds: ")
+                }
+            }
+
+        }
+    }
+
+    @Test
+    fun comparableHell() {
+        inlineModelTest(
+            """
+                |/src/main/kotlin/function/Test.kt
+                |package comparable
+                | /**
+                | * generic function
+                | */
+                |public fun <T : Comparable<R>, R> generic() {}
+                |class Test<T>() {}
+        """
+        ) {
+            with((this / "comparable" / "generic").assertIsInstance<Function>("generic")) {
+                name.assertEqual("generic", "Function name: ")
+                visibility.values.first().assertEqual(org.jetbrains.kotlin.descriptors.Visibilities.PUBLIC)
+                parameters.assertCount(0)
+                returnType.assertNotNull("returnType").constructorFqName.assertEqual("kotlin.Unit", "Return type: ")
+
+                typeParameters.assertCount(2, "Type parameters: ")
+                val T = typeParameters.find { it.name == "T" }.assertIsInstance<GenericType>("T")
+                val R = typeParameters.find { it.name == "R" }.assertIsInstance<GenericType>("R")
+
+                with(T) {
+                    name.assertEqual("T", "First type parameter name: ")
+                    bounds.assertCount(1, "First type parameter bounds: ")
+                    with(bounds.single().assertIsInstance<TypeConstructor>("Comparable")) {
+                        name.assertEqual("kotlin.Comparable", "Comparable TypeConstructor: ")
+                        bounds.assertCount(1, "Comparable bounds: ")
+                        bounds.single().name.assertEqual("R", "Comparable bound: ")
```

Comment:
Comparable has no type bounds, it only has a type parameter (which is `R` in this case)